src/helpers/gateway/mod.rs

```diff
@@ -116,6 +138,85 @@ impl<T: Transport> Gateway<T> {
                 .get_or_create(channel_id, || self.transport.receive(channel_id)),
         )
     }
+
+    fn create_idle_tracker(senders: Arc<GatewaySenders>, receivers: Arc<GatewayReceivers<T>>) -> tokio::task::JoinHandle<()> {
+        tokio::task::spawn(async move {
+            // Perform some periodic work in the background
+            loop {
+                let _ = tokio::time::sleep(Duration::from_secs(5)).await;
+                if senders.check_idle_and_reset() && receivers.check_idle_and_reset() {
+                    let sender_missing_records =senders.get_all_missing_records();
+                    let sender_message : HashMap<&ChannelId, String> = sender_missing_records.iter().map(|entry|{
+                        let (SenderStatus {next, current_written , buf_size, message_size},missing_records) = entry.1;
+                        let last_pending_message = missing_records.iter().cloned().max();
+                            let last_pending_message = last_pending_message.unwrap();
+                            let chunk_head = next - current_written / message_size;
+                            let chunk_size = buf_size / message_size;
+                            let chunk_count = (last_pending_message - chunk_head + chunk_size - 1) / chunk_size;
+                            let mut response = String::new();
+                            for i in 0..chunk_count {
+                                let chunk_response_head = format!("The next, {}-th chunk, missing: ", i);
+                                let mut chunk_response = Vec::new();
+                                for j in (chunk_head + i * chunk_size).. chunk_head + (i+1)* chunk_size {
+                                    if ! missing_records.contains(&j) {
+                                        chunk_response.push(j);
+                                    }
+                                }
+                                if chunk_response.len() > 0 {
+                                    response = response + &chunk_response_head + &Self::compress_numbers(&chunk_response) + " if not closed early.\n";
```

Comment:
it is a bit hard to read - would it make sense to do this

* ask sender buffer to give us all records ids that are missing
* convert `Vec<usize>` to `Vec<MissingRange<usize>>` using the contiguous range algo
* spit out `Vec<MissingRange<usize>>` to logging system using the debug formatter `{:?}'

and the missing range is just a wrapper around standard `Range` struct to allow custom formatting

```rust
struct MissingRange(Range<usize>);

impl Debug for MissingRange {
   .. // format start..end as [start..end] 
      // if start == end format as [start]
}
```

the more stuff we pull out of this function, the easier it will be to understand what is going on here